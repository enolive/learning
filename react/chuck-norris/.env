REACT_APP_API_URL=http://myserver/api
