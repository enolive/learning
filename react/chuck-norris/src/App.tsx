import React, {useEffect, useReducer} from 'react';
import logo from './logo.jpg';
import './App.css';

export interface Joke {
  value: {
    joke: string
  }
}

export const tidy = (text: string) =>
  text.replace(/&quot;/g, '"');

type State =
  | { status: 'empty' }
  | { status: 'loading' }
  | { status: 'success', jokeText: string }
  | { status: 'failure', error: string }

type Action =
  | { type: 'fetch' }
  | { type: 'success', result: string }
  | { type: 'failure', error: string };

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case 'success':
      return {status: 'success', jokeText: action.result};
    case 'failure':
      return {status: 'failure', error: action.error};
    case 'fetch':
      return {status: 'loading'};
  }
};

function App() {
  const [jokeData, dispatch] = useReducer(reducer, {status: 'empty'});
  const fetchRandomJoke = async () => {
    dispatch({type: 'fetch'});
    await fetch('http://api.icndb.com/jokes/random')
      .then(response => response.json() as Promise<Joke>)
      .then(joke => joke.value.joke)
      .then(tidy)
      .then(joke => dispatch({type: 'success', result: joke}))
      .catch((e: Error) => dispatch({type: 'failure', error: e.message}));
  };

  useEffect(() => {
    fetchRandomJoke();
  }, []);

  const refreshJoke = async () => {
    await fetchRandomJoke();
  };

  return (
    <div className="App">
      <div className="container-fluid">
        <header className="row">
          <figure className="figure">
            <img className="figure-caption" src={logo} alt="Chuck Norris Logo"/>
            <figcaption className="figure-caption">
              <a data-testid="attribution"
                 href="https://www.flickr.com/photos/enriquespics/2058486062">
                by Carlos Killpack
              </a>
            </figcaption>
          </figure>
        </header>
        <main className="row">
          {jokeData.status === 'success' && <p data-testid="joke-text">Random Wisdom: {jokeData.jokeText}.</p>}
          {jokeData.status === 'failure' && <p data-testid="error">{jokeData.error}</p>}
        </main>
        <footer className="row">
          <button className="btn btn-primary"
                  onClick={refreshJoke}
                  disabled={jokeData.status === 'loading'}>
            Refresh
          </button>
        </footer>
      </div>
    </div>
  );
}

export default App;
