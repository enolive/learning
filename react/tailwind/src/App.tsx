import React, {FC} from 'react'

const App: FC = () => (
    <div className="h-screen bg-amber-600">
        <h1 className="m-6 text-amber-50 text-5xl bg-amber-800 rounded">Welcome to my App!</h1>
        <p className="font-extrabold">
        just a simple test for tailwind
        </p>
    </div>
)

export default App
