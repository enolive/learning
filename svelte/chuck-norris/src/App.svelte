<script lang="ts">
  interface JokeText {
    value: string;
  }

  function fetchRandomJoke() {
    fetchJoke$ = fetch('https://api.chucknorris.io/jokes/random')
      .then(res => res.json() as JokeText)
      .then(joke => joke.value);
  }

  let fetchJoke$: Promise<string>;

  fetchRandomJoke();
</script>

<header>
  <h1>Chuck Norris</h1>
</header>
<main>
  <figure>
    <img src="chuck-norris-logo.png" alt="Chuck Norris approves!">
    <figcaption>Chuck Norris approves!</figcaption>
  </figure>
  <p role="status">
    {#await fetchJoke$}
      Loading...
    {:then jokeText}
      Random Wisdom: {jokeText}
    {/await}
  </p>
  <button on:click={fetchRandomJoke}>Refresh</button>
</main>
