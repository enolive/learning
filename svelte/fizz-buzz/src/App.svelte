<script lang="ts">
  import {fizzBuzz} from './fizzbuzz';

  let input: number;
  let result: string;

  const calculateIt = () => {
    result = fizzBuzz(input);
  };
</script>

<main>
    <h1>Fizz Buzz</h1>
    <form on:submit|preventDefault={calculateIt}>
        <label>
            Enter a number:
            <input type="number" bind:value={input}>
        </label>
        <button>Calculate It</button>
        {#if result}
          <div role="status">Result: {result}</div>
        {/if}
    </form>
</main>

<style>
    main {
        text-align: center;
        padding: 1em;
        max-width: 240px;
        margin: 0 auto;
    }

    h1 {
        color: #ff3e00;
        text-transform: uppercase;
        font-size: 4em;
        font-weight: 100;
    }

    @media (min-width: 640px) {
        main {
            max-width: none;
        }
    }
</style>
