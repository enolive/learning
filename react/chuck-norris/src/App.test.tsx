import React from 'react';
import {fireEvent, render, waitForDomChange} from '@testing-library/react';
import App, {reducer, State, tidy} from './App';

const jokeResponse = (text: string = 'whatever') => (
  `{
    "type": "success",
    "value": {
      "id": 94,
      "joke": "${text}",
      "categories": []
    }
  }
`);

beforeEach(() => fetchMock.resetMocks());

describe('tidy', () => {
  test('tidy replaces quotation html entities', () => {
    expect(tidy('&quot;&quot;')).toBe('""');
  });
});

describe('reducer', () => {
  const previousStates: State[] = [
    {status: 'empty'},
    {status: 'failure', error: 'some error we care not about'},
    {status: 'loading'},
    {status: 'success', jokeText: 'old joke'}
  ];

  describe.each(previousStates)('previous state %j', (previous: State) => {
    test('fetch triggers loading state', () => {
      const result = reducer(previous, {type: 'fetch'});
      expect(result).toEqual({status: 'loading'});
    });

    test('failure triggers error state', () => {
      const result = reducer(previous, {type: 'failure', error: 'something went wrong'});
      expect(result).toEqual({status: 'failure', error: 'something went wrong'});
    });

    test('success triggers result state', () => {
      const result = reducer(previous, {type: 'success', result: 'a joke!'});
      expect(result).toEqual({status: 'success', jokeText: 'a joke!'});
    });
  });
});


describe('rendering', () => {
  test('renders chuck norris image giving attribution to the author', () => {
    const {getByRole, getByTestId} = render(<App/>);

    const image = getByRole('img');
    const attribution = getByTestId('attribution');
    expect(image).toHaveAttribute('src', 'logo.jpg');
    expect(image).toHaveAttribute('alt', 'Chuck Norris Logo');
    expect(attribution).toMatchSnapshot();
  });

  test('renders errors', async () => {
    fetchMock.mockRejectOnce(new Error('I AM ERROR'));

    const {getByTestId, queryByRole} = render(<App/>);
    await waitForDomChange();

    const error = getByTestId('error');
    expect(error).toHaveTextContent('I AM ERROR');
    expect(queryByRole('status')).toBeNull();
  });

  test('renders the retrieved joke', async () => {
    fetchMock.mockResponseOnce(jokeResponse('there\'s no place like 127.0.0.1&quot;'));
    const {getByTestId, queryByTestId, queryByRole} = render(<App/>);
    const jokeBefore = queryByTestId('joke-text');
    expect(jokeBefore).toBeNull();
    await waitForDomChange();

    const joke = getByTestId('joke-text');
    expect(joke).toHaveTextContent('Random Wisdom: there\'s no place like 127.0.0.1".');
    expect(queryByRole('status')).toBeNull();
  });

  test('fetches a random joke from the web', async () => {
    fetchMock.mockResponseOnce(jokeResponse());
    render(<App/>);

    expect(fetchMock.mock.calls.length).toBe(1);
    expect(fetchMock.mock.calls[0][0]).toBe('http://api.icndb.com/jokes/random');
  });
});

describe('refresh', () => {
  test('allows refreshing', async () => {
    fetchMock.mockResponseOnce(jokeResponse());
    const {getByRole} = render(<App/>);
    await waitForDomChange();

    const refresh = getByRole('button');
    expect(refresh).toHaveTextContent('Refresh');
  });

  test('disallows refreshing while refresh is in work', async () => {
    const {getByRole} = render(<App/>);

    const refresh = getByRole('button');
    expect(refresh).toHaveAttribute('disabled', '');
  });

  test('shows loading status while refresh is in work', async () => {
    const {getByRole} = render(<App/>);

    const loading = getByRole('status');
    expect(loading).toHaveTextContent('Loading...');
  });

  test('refreshing gets a new joke', async () => {
    fetchMock.mockResponseOnce(jokeResponse());
    const {getByRole} = render(<App/>);
    await waitForDomChange();

    fireEvent.click(getByRole('button'));

    expect(fetchMock.mock.calls.length).toBe(2);
    expect(fetchMock.mock.calls[1][0]).toBe('http://api.icndb.com/jokes/random');
  });

  test('refreshing re-renders the joke text', async () => {
    fetchMock.mockResponseOnce(jokeResponse());
    fetchMock.mockResponseOnce(jokeResponse('second'));
    const {getByRole, getByTestId} = render(<App/>);
    await waitForDomChange();

    fireEvent.click(getByRole('button'));
    await waitForDomChange();

    expect(getByTestId('joke-text')).toHaveTextContent('second');
  });
});
