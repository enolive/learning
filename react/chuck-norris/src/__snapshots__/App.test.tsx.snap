// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders chuck norris image giving attribution to the author 1`] = `
<a
  data-testid="attribution"
  href="https://www.flickr.com/photos/enriquespics/2058486062"
>
  by Carlos Killpack
</a>
`;
